src/basic/MAT_MAT_SHARED-Cuda.cpp

```diff
@@ -0,0 +1,279 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "MAT_MAT_SHARED.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#if defined(RAJA_ENABLE_CUDA)
+
+#include "common/CudaDataUtils.hpp"
+
+#include <iostream>
+
+namespace rajaperf {
+namespace basic {
+
+#define MAT_MAT_SHARED_DATA_SETUP_CUDA                                         \
+  const Index_type NN = m_N * m_N;                                             \
+  allocAndInitCudaDeviceData(A, m_A, NN);                                      \
+  allocAndInitCudaDeviceData(B, m_B, NN);                                      \
+  allocAndInitCudaDeviceData(C, m_C, NN);
+
+#define MAT_MAT_SHARED_DATA_TEARDOWN_CUDA                                      \
+  getCudaDeviceData(m_A, A, NN);                                               \
+  getCudaDeviceData(m_B, B, NN);                                               \
+  getCudaDeviceData(m_C, C, NN);                                               \
+  deallocCudaDeviceData(A);                                                    \
+  deallocCudaDeviceData(B);                                                    \
+  deallocCudaDeviceData(C);
+
+__global__ void mat_mat_shared(Index_type N, Real_ptr C, Real_ptr A,
```

Comment:
will this be able to execute properly for arbitrary problem sizes?

Reply:
It should N is the height of the matrix, one issue that came up though is we can reach overflow in the flops/rep. Going over 5million lead to this issue. Should I try to fix that here.

Reply:
Fixing flops per rep should be easy if you use Index_type instead of int, go ahead.

Reply:
Does this work for N that do not evenly divide the tile size?